lib/Doctrine/ORM/PersistentCollection.php

```diff
@@ -692,23 +692,41 @@ public function unwrap()
     protected function doInitialize()
     {
         // Has NEW objects added through add(). Remember them.
-        $newObjects = [];
+        $newlyAddedDirtyObjects = [];
 
         if ($this->isDirty) {
-            $newObjects = $this->collection->toArray();
+            $newlyAddedDirtyObjects = $this->collection->toArray();
         }
 
         $this->collection->clear();
```

Comment:
I think the issue could be solved by only not calling `clear()` here. 
IIRC, `UnitOfWork` already handles dirty elements in the collection and does not replace them when hydrating in `BasicEntityPersister::loadCollectionFromStatement` (which is where all calls go at the end).